File: app/src/main/java/rbsoftware/friendstagram/service/UsersService.java
package rbsoftware.friendstagram.service;

import java.util.Map;

import rbsoftware.friendstagram.model.Response;
import rbsoftware.friendstagram.model.User;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;

/**
 * Created by silver_android on 1/5/2017.
 */

public class UsersService extends NetworkService {

    private final UserAPI api;
    private static UsersService instance;

    private UsersService() {
        api = getRetrofit().create(UserAPI.class);
    }

    public static UsersService getInstance() {
        if (instance == null)
            instance = new UsersService();
        return instance;
    }

    public UserAPI getAPI() {
        return api;
    }

    public interface UserAPI {
        @POST("/users")
        Call<Response<User>> register(@Body Map<String, String> registerJSON);

        @POST("/users/login")
        Call<Response<String>> login(@Body Map<String, String> loginJSON);

        @GET("/users/{username}")
        Call<Response<User>> getUser(@Path("username") String username);

        @POST("/users/logoff")
        Call<User> logout();
    }
}

File: app/src/main/java/rbsoftware/friendstagram/service/UsersService.kt
package rbsoftware.friendstagram.service

import io.reactivex.Single
import rbsoftware.friendstagram.model.LoginResponse
import rbsoftware.friendstagram.model.ServerResponse
import rbsoftware.friendstagram.model.User
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.http.*
import javax.inject.Inject

/**
 * Created by silver_android on 1/5/2017.
 */

class UsersService @Inject constructor(retrofit: Retrofit) {
    val api: UserAPI

    init {
        api = retrofit.create(UserAPI::class.java)
    }

    interface UserAPI {
        @POST("/users")
        fun register(@Body registerJSON: Map<String, String>): Single<Response<ServerResponse<User>>>

        @POST("/users/login")
        fun login(@Body loginJSON: Map<String, String>): Single<Response<ServerResponse<LoginResponse>>>

        @GET("/users/{username}")
        fun getUser(@Path("username") username: String): Single<Response<ServerResponse<User>>>

        @PUT("/users/{username}")
        fun updateUser(@Path("username") username: String, @Body updateJSON: Map<String, @JvmSuppressWildcards Any>): Single<Response<ServerResponse<String>>>

        @PUT("/users/profile_picture")
        fun updateUserProfilePicture(@Body imageURL: Map<String, String>): Single<Response<ServerResponse<String>>>

        @POST("/users/logoff")
        fun logout(): Single<User>
    }
}
